### PR TITLE
CoreData implementation - Maksim Soldatov

### PR DESCRIPTION
**Pull request by MaksimSoldatov**

### CODE REVIEW

File: FeedStoreChallenge/CoreData/CoreDataFeedStore.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    private let context: NSManagedObjectContext
+
+    public init(storeURL: URL, bundle: Bundle = .main) throws {
+        container = try NSPersistentContainer.load(modelName: "FeedStore", url: storeURL, in: bundle)
+        context = container.newBackgroundContext()
+    }
+    
+    public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+        perform { context in
+            try! ManagedCache.find(in: context).map(context.delete).map(context.save)
```
Comment (caiozullo) [inline FeedStoreChallenge/CoreData/CoreDataFeedStore.swift:23]: `try!` may lead to runtime errors - use `do/catch` instead to handle errors. If you catch an error, complete with the error.

File: FeedStoreChallenge/CoreData/CoreDataFeedStore.swift [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    
+    public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+        perform { context in
+            
+            let managedCache = try! ManagedCache.newUniqueInstance(in: context)
+            managedCache.timestamp = timestamp
+            managedCache.feed = ManagedFeedImage.images(from: feed, in: context)
+            
+            try! context.save()
```
Comment (caiozullo) [inline FeedStoreChallenge/CoreData/CoreDataFeedStore.swift:35]: `try!` may lead to runtime errors - use `do/catch` instead to handle errors. If you catch an error, complete with the error.

File: FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import CoreData
+
+internal extension NSPersistentContainer {
+    internal enum LoadingError: Swift.Error {
+        case modelNotFound
+        case failedToLoadPersistentStores(Swift.Error)
+    }
+    
+    internal static func load(modelName name: String, url: URL, in bundle: Bundle) throws -> NSPersistentContainer {
```
Comment (caiozullo) [inline FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift:17]: `internal` is redundant in this scope - it generates a warning. You can remove it.

File: FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+//  FeedStoreChallenge
+//
+//  Created by Maxim Soldatov on 10/29/20.
+//  Copyright © 2020 Essential Developer. All rights reserved.
+//
+
+import CoreData
+
+internal extension NSPersistentContainer {
+    internal enum LoadingError: Swift.Error {
```
Comment (caiozullo) [inline FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift:12]: `internal` is redundant in this scope - it generates a warning. You can remove it.

**Review by caiozullo — COMMENTED** [review]
Hi @MaximSoldier,

You almost passed the challenge! 

There are unhandled failure paths that can lead to runtime errors and some warnings in the project. I left some comments in the review.

Feel free to fix it and push your changes. Thank you ✅

File: FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift (hunk 3 above)
Comment (MaximSoldier) [reply FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift]: Its looks fixed in this commit: 49ca8401435b98230c4db7cd24f57d053e620c19
Remove redundant internal access controls

File: FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift (hunk 4 above)
Comment (MaximSoldier) [reply FeedStoreChallenge/CoreData/CoreDataStoreHelper.swift]: Its looks fixed here: 49ca8401435b98230c4db7cd24f57d053e620c19

File: FeedStoreChallenge/CoreData/CoreDataFeedStore.swift (hunk 2 above)
Comment (MaximSoldier) [reply FeedStoreChallenge/CoreData/CoreDataFeedStore.swift]: Fexed: 76c3c94b4beba7bfa3a9a7dd0cd5d404bc1efd29

File: FeedStoreChallenge/CoreData/CoreDataFeedStore.swift (hunk 1 above)
Comment (MaximSoldier) [reply FeedStoreChallenge/CoreData/CoreDataFeedStore.swift]: Fexed: 76c3c94b4beba7bfa3a9a7dd0cd5d404bc1efd29

**Review by caiozullo — COMMENTED** [review]
Hi @MaximSoldier,

Congratulations, you've passed the challenge! 🥇

- All tests are passing ✅ 
- You broke down the implementation into small commits ✅ 
- Bonus points for the Integration Tests and Memory Leak Tracking ✅ 

Great job!

**Pull request closed without merging**